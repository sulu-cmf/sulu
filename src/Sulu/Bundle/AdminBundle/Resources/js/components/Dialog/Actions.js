// @flow
import React from 'react';
import type {Action} from './types';

type Props = {
    actions: Array<Action>,
};

export default class Actions extends React.PureComponent<Props> {
    render() {
        const {actions} = this.props;
        if (!actions.length) {
            return null;
        }

        return (
            <div>
                {actions.map((action, index) => {
                    const handleButtonClick = action.onClick;
                    return (
                        <button
                            key={index}
                            onClick={handleButtonClick}
                        >
                            {action.title}
                        </button>
                    );
                })}
            </div>
        );
    }
}
