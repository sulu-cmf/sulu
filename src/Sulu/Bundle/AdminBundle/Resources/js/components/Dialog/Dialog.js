// @flow
import classNames from 'classnames';
import {observable, action} from 'mobx';
import {observer} from 'mobx-react';
import React from 'react';
import type {Node} from 'react';
import {Portal} from 'react-portal';
import {afterElementsRendered} from '../../services/DOM';
import Backdrop from '../Backdrop';
import Button from '../Button';
import dialogStyles from './dialog.scss';

type Props = {|
    confirmDisabled: boolean,
    children: Node,
    cancelText: string,
    confirmText: string,
    confirmLoading: boolean,
    onConfirm: () => void,
    onCancel: () => void,
    open: boolean,
    size?: 'small' | 'large',
    title: string,
|};

@observer
export default class Dialog extends React.Component<Props> {
    static defaultProps = {
        confirmDisabled: false,
        confirmLoading: false,
        open: false,
    };

    @observable visible: boolean = false;
    @observable openHasChanged: boolean = false;

    constructor(props: Props) {
        super(props);

        this.openHasChanged = this.props.open;
    }

    componentDidMount() {
        this.toggle();
    }

    @action componentWillReceiveProps(newProps: Props) {
        this.openHasChanged = newProps.open !== this.props.open;
    }

    componentDidUpdate() {
        this.toggle();
    }

    close = () => {
        this.props.onCancel();
    };

    @action toggle() {
        afterElementsRendered(action(() => {
            if (this.openHasChanged) {
                this.visible = this.props.open;
            }
        }));
    }

    @action handleTransitionEnd = () => {
        afterElementsRendered(action(() => {
            this.openHasChanged = false;
        }));
    };

    render() {
        const {
            open,
            title,
            children,
            confirmDisabled,
            onCancel,
            onConfirm,
            cancelText,
            confirmText,
            confirmLoading,
            size,
        } = this.props;
        const containerClass = classNames(
            dialogStyles.dialogContainer,
            {
                [dialogStyles.open]: this.visible,
            }
        );

        const showPortal = open || this.openHasChanged;

        const dialogClass = classNames(
            dialogStyles.dialog,
            {
                [dialogStyles[size]]: size,
            }
        );

        return (
            <div>
                <Backdrop open={showPortal} />
                {showPortal &&
                    <Portal>
                        <div
                            className={containerClass}
                            onTransitionEnd={this.handleTransitionEnd}
                        >
                            <div className={dialogClass}>
                                <section className={dialogStyles.content}>
                                    <header>
                                        {title}
                                    </header>
                                    <article>
                                        {children}
                                    </article>
                                    <footer>
                                        <Button onClick={onCancel} skin="secondary">
                                            {cancelText}
                                        </Button>
                                        <Button
                                            disabled={confirmDisabled}
                                            loading={confirmLoading}
                                            onClick={onConfirm}
                                            skin="primary"
                                        >
                                            {confirmText}
                                        </Button>
                                    </footer>
                                </section>
                            </div>
                        </div>
                    </Portal>
                }
            </div>
        );
    }
}
