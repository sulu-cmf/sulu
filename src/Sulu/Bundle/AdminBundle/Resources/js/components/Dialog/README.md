The `Dialog` component let's you display some content above everything else.
It renders depending on the passed property and request being closed through a callback.

```
initialState = {open: false};

const onConfirm = () => {
    /* do confirm things */
    setState({open: false});
};

const onCancel = () => {
    /* do cancel things */
    setState({open: false});
};

<div>
    <button onClick={() => setState({open: true})}>Open dialog</button>
    <Dialog
        title="Question?"
        onCancel={onCancel}
        onConfirm={onConfirm}
        cancelText="No"
        confirmText="Yes"
        open={state.open}>
        You've got a question in here.
        Yes or no?
    </Dialog>
</div>
```
