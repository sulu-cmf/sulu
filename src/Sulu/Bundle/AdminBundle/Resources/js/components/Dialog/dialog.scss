@import '../../containers/Application/colors.scss';

$contentBackgroundColor: $concrete;
$footerBackgroundColor: $white;
$borderRadius: 3px;

$titleColor: $black;
$contentColor: $dustyGray;

$confirmBackground: $shakespeare;
$confirmColor: $white;

$transitionDuration: 300ms;

.dialog-container {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 100;
}

.dialog {
    width: 500px;
    position: relative;
    opacity: 0;
    will-change: opacity;
    transition: opacity $transitionDuration;
    z-index: 1;
    pointer-events: none;

    & > * {
        pointer-events: auto;
    }
}

.open {
    & > * {
        opacity: 1;
    }
}

.content {
    border-radius: $borderRadius;
    background: $contentBackgroundColor;
    overflow: hidden;

    header {
        color: $titleColor;
        font-size: 22px;
        font-weight: bold;
        text-align: center;
        padding: 30px;
    }

    article {
        color: $contentColor;
        overflow: auto;
        padding: 0 30px 30px;
        text-align: center;
    }

    footer {
        background: $footerBackgroundColor;
        padding: 30px;
        display: flex;
        justify-content: space-between;
    }
}
