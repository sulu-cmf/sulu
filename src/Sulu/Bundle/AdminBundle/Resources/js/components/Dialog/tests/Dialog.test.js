// @flow
import {mount, shallow} from 'enzyme';
import React from 'react';
import pretty from 'pretty';
import Dialog from '../Dialog';

afterEach(() => {
    if (document.body) {
        document.body.innerHTML = '';
    }
});

test('The component should render in body when open', () => {
    const body = document.body;
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const view = mount(
        <Dialog
            cancelText="Cancel"
            confirmText="Confirm"
            onCancel={onCancel}
            onConfirm={onConfirm}
            open={true}
            title="My dialog title"
        >
            <div>My dialog content</div>
        </Dialog>
    ).render();

    expect(view).toMatchSnapshot();
    expect(pretty(body ? body.innerHTML : '')).toMatchSnapshot();
});

test('The component should render in body with disabled confirm button', () => {
    const body = document.body;
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const view = mount(
        <Dialog
            cancelText="Cancel"
            confirmDisabled={true}
            confirmText="Confirm"
            onCancel={onCancel}
            onConfirm={onConfirm}
            open={true}
            title="My dialog title"
        >
            <div>My dialog content</div>
        </Dialog>
    ).render();

    expect(view).toMatchSnapshot();
    expect(pretty(body ? body.innerHTML : '')).toMatchSnapshot();
});

test('The component should render in body with a large class', () => {
    const body = document.body;
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const view = mount(
        <Dialog
            cancelText="Cancel"
            confirmText="Confirm"
            onCancel={onCancel}
            onConfirm={onConfirm}
            open={true}
            size="large"
            title="My dialog title"
        >
            <div>My dialog content</div>
        </Dialog>
    ).render();

    expect(view).toMatchSnapshot();
    expect(pretty(body ? body.innerHTML : '')).toMatchSnapshot();
});

test('The component should render in body with loader instead of confirm button', () => {
    const body = document.body;
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const view = mount(
        <Dialog
            cancelText="Cancel"
            confirmLoading={true}
            confirmText="Confirm"
            onCancel={onCancel}
            onConfirm={onConfirm}
            open={true}
            title="My dialog title"
        >
            <div>My dialog content</div>
        </Dialog>
    ).render();

    expect(view).toMatchSnapshot();
    expect(pretty(body ? body.innerHTML : '')).toMatchSnapshot();
});

test('The component should not render in body when closed', () => {
    const body = document.body;
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const view = mount(
        <Dialog
            cancelText="Cancel"
            confirmText="Confirm"
            onCancel={onCancel}
            onConfirm={onConfirm}
            open={false}
            title="My dialog title"
        >
            My dialog content
        </Dialog>
    ).render();

    expect(view).toMatchSnapshot();
    expect(body ? body.innerHTML : '').toBe('');
});

test('The component should call the callback when the confirm button is clicked', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const view = shallow(
        <Dialog
            cancelText="Cancel"
            confirmText="Confirm"
            onCancel={onCancel}
            onConfirm={onConfirm}
            open={true}
            title="My dialog title"
        >
            My dialog content
        </Dialog>
    );

    expect(onConfirm).not.toBeCalled();
    view.find('Button[skin="primary"]').simulate('click');
    expect(onConfirm).toBeCalled();
});

test('The component should call the callback when the cancel button is clicked', () => {
    const onConfirm = jest.fn();
    const onCancel = jest.fn();
    const view = shallow(
        <Dialog
            cancelText="Cancel"
            confirmText="Confirm"
            onCancel={onCancel}
            onConfirm={onConfirm}
            open={true}
            title="My dialog title"
        >
            My dialog content
        </Dialog>
    );

    expect(onCancel).not.toBeCalled();
    view.find('Button[skin="secondary"]').simulate('click');
    expect(onCancel).toBeCalled();
});
