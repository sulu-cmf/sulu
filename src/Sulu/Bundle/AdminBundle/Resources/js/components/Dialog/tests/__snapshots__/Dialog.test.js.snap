// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should not render in body when closed 1`] = `<div />`;

exports[`The component should render in body when open 1`] = `<div />`;

exports[`The component should render in body when open 2`] = `
"<div>
  <div class=\\"backdrop visible fixed\\"></div>
</div>
<div>
  <div class=\\"dialogContainer\\">
    <div class=\\"dialog\\">
      <section class=\\"content\\">
        <header>My dialog title</header>
        <article>
          <div>My dialog content</div>
        </article>
        <footer><button class=\\"button secondary\\" type=\\"button\\"><span class=\\"text\\">Cancel</span></button><button class=\\"button primary\\" type=\\"button\\"><span class=\\"text\\">Confirm</span></button></footer>
      </section>
    </div>
  </div>
</div>"
`;

exports[`The component should render in body with a large class 1`] = `<div />`;

exports[`The component should render in body with a large class 2`] = `
"<div>
  <div class=\\"backdrop visible fixed\\"></div>
</div>
<div>
  <div class=\\"dialogContainer\\">
    <div class=\\"dialog large\\">
      <section class=\\"content\\">
        <header>My dialog title</header>
        <article>
          <div>My dialog content</div>
        </article>
        <footer><button class=\\"button secondary\\" type=\\"button\\"><span class=\\"text\\">Cancel</span></button><button class=\\"button primary\\" type=\\"button\\"><span class=\\"text\\">Confirm</span></button></footer>
      </section>
    </div>
  </div>
</div>"
`;

exports[`The component should render in body with disabled confirm button 1`] = `<div />`;

exports[`The component should render in body with disabled confirm button 2`] = `
"<div>
  <div class=\\"backdrop visible fixed\\"></div>
</div>
<div>
  <div class=\\"dialogContainer\\">
    <div class=\\"dialog\\">
      <section class=\\"content\\">
        <header>My dialog title</header>
        <article>
          <div>My dialog content</div>
        </article>
        <footer><button class=\\"button secondary\\" type=\\"button\\"><span class=\\"text\\">Cancel</span></button><button class=\\"button primary\\" disabled=\\"\\" type=\\"button\\"><span class=\\"text\\">Confirm</span></button></footer>
      </section>
    </div>
  </div>
</div>"
`;

exports[`The component should render in body with loader instead of confirm button 1`] = `<div />`;

exports[`The component should render in body with loader instead of confirm button 2`] = `
"<div>
  <div class=\\"backdrop visible fixed\\"></div>
</div>
<div>
  <div class=\\"dialogContainer\\">
    <div class=\\"dialog\\">
      <section class=\\"content\\">
        <header>My dialog title</header>
        <article>
          <div>My dialog content</div>
        </article>
        <footer><button class=\\"button secondary\\" type=\\"button\\"><span class=\\"text\\">Cancel</span></button><button class=\\"button primary loading\\" disabled=\\"\\" type=\\"button\\"><span class=\\"text\\">Confirm</span>
            <div class=\\"loader\\">
              <div class=\\"spinner\\" style=\\"width: 25px; height: 25px;\\">
                <div class=\\"doubleBounce1\\"></div>
                <div class=\\"doubleBounce2\\"></div>
              </div>
            </div></button></footer>
      </section>
    </div>
  </div>
</div>"
`;
